### PR TITLE
Use state machine for web REPL

### PR DESCRIPTION
**Pull request by brian-carroll**
The CLI REPL uses a state machine to allow the user to enter definitions with no final expression. Definitions are type checked, but no code is generated or executed until the user enters an expression.

This PR updates the Web REPL to share the same state machine, so that it will behave in the same way.

This involved some refactoring because previously the state machine made direct calls into the LLVM back end. The Web REPL is built _without_ an LLVM back end, only the Wasm development back end.

As of this PR, the state machine does not call any back end, but instead returns a `ReplAction` enum, telling the caller whether it should evaluate the user program. It still executes all compiler phases up to monomorphization, so that we can share error reporting code.

Screenshot 2023-09-09 at 16 52 42 (2)

### CODE REVIEW

**Comment by brian-carroll** [comment]
There seems to be an issue with sending multi-line input to the state machine

Screenshot 2023-09-09 at 17 22 42


But I would love to get this PR merged and then follow up with fixes afterwards!

**Comment by rtfeldman** [comment]
I'm cool with that plan! Looks like test failures are web repl tests at the moment.

**Comment by brian-carroll** [comment]
Oops! That's fixed now.
Previously, the web REPL only had colour for the `»` and `…` prompts, and red for compiler panics. But now there are HTML tags in the actual result string, so we have to strip them in tests.

File: crates/repl_ui/src/repl_state.rs [hunk 1]
```diff
@@ -82,39 +34,56 @@ impl Default for ReplState {
...
-        match parse_src(&arena, line) {
-            ParseOutcome::Empty => Ok(TIPS.to_string()),
+    pub fn step<'a>(
+        &mut self,
+        arena: &'a Bump,
+        line: &str,
+        target_info: TargetInfo,
+        palette: Palette,
+    ) -> ReplAction<'a> {
+        match parse_src(arena, line) {
```
Comment (brian-carroll) [inline crates/repl_ui/src/repl_state.rs:66]: @rtfeldman I noticed that we seem to be parsing here, then descending into `next_action`, where we parse again! This doesn't seem necessary. Is that a mistake, or am I missing some subtlety?
If it's a subtlety then we should probably explain it in a comment.

Comment (brian-carroll) [reply crates/repl_ui/src/repl_state.rs]: I refactored it to get rid of this, and it still seems to pass Wasm tests locally.
(I can't run the other REPL tests due to a problem I'm having with `-lSystem`)

Comment (rtfeldman) [reply crates/repl_ui/src/repl_state.rs]: Yeah I think you're right, there was a redundant parse in there!

**Review by rtfeldman — APPROVED** [review]
Excellent, thanks @brian-carroll! 😻

**Pull request merged**